src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -50,16 +52,54 @@ universes u v
 -- The domain of our additive characters
 variables (R : Type u) [add_monoid R]
 -- The target
-variables (R' : Type v) [monoid R']
+variables (R' : Type v) [comm_monoid R']
 
 /-- Define `add_char R R'` as `(multiplicative R) →* R'`.
 The definition works for an additive monoid `R` and a monoid `R'`,
 but we will restrict to the case that both are commutative rings below.
+We assume right away that `R'` is commutative, so that `add_char R R'` carries
+a structure of commutative monoid.
 The trivial additive character (sending everything to `1`) is `(1 : add_char R R').` -/
-abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+@[derive [comm_monoid, inhabited]]
+def add_char : Type (max u v) := (multiplicative R) →* R'
+
+instance add_char.has_coe_to_fun : has_coe_to_fun (add_char R R') (λ x, (multiplicative R → R')) :=
+monoid_hom.has_coe_to_fun
+
+instance add_char.monoid_hom_class : monoid_hom_class (add_char R R') (multiplicative R) R' :=
+monoid_hom.monoid_hom_class
 
 end add_char_def
 
+section group_structure
+
+universes u v
+
+namespace add_char
+
+open multiplicative
+
+variables {R : Type u} [add_comm_group R] {R' : Type v} [comm_monoid R']
+
+/-- An additive character on a commutative additive group has an inverse. -/
+instance has_inv : has_inv (add_char R R') := ⟨λ ψ, ψ.comp neg_add_monoid_hom.to_multiplicative⟩
```

Comment:
Does this work as
```suggestion
instance has_inv : has_inv (add_char R R') := ⟨λ ψ, ψ.comp inv_monoid_hom⟩
```

Reply:
OK. I was thinking in too complicated ways...